### PR TITLE
Add Strict Connection Manager

### PR DESCRIPTION
**Pull request by nisdas**
- [x] We add in a forked connection manager with a much smaller period in which the background connection checker loop runs in. 
- [x] Protect bootnode/relay nodes from pruning.

### CODE REVIEW

File: beacon-chain/p2p/connmgr.go [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+
+	firstSeen time.Time // timestamp when we began tracking this peer.
+}
+
+// TrimOpenConns closes the connections of as many peers as needed to make the peer count
+// equal the low watermark. Peers are sorted in ascending order based on their total value,
+// pruning those peers with the lowest scores first, as long as they are not within their
+// grace period.
+//
+// TODO: error return value so we can cleanly signal we are aborting because:
```
Comment (prestonvanloon) [inline beacon-chain/p2p/connmgr.go]: This is failing lint.

**Comment by nisdas** [comment]
Closing this in favour of #4133

**Comment by nisdas** [comment]
Reopening this to experiment with a strict connection manager

File: beacon-chain/p2p/handshake.go [hunk 2]
```diff
@@ -41,7 +41,10 @@ func (s *Service) AddConnectionHandler(reqFunc func(ctx context.Context, id peer
 				if err := reqFunc(ctx, conn.RemotePeer()); err != nil && err != io.EOF {
 					log.WithError(err).Debug("Could not send successful hello rpc request")
 					if err.Error() == "protocol not supported" {
+						// This is only to ensure the smooth running of our testnets. This will not be
```
Comment (prestonvanloon) [inline beacon-chain/p2p/handshake.go]: yeah.. gotta think of a better solution for mainnet :sweat_smile:

**Comment by prestonvanloon** [comment]
Looks OK to me. I'm worried about the trim loop and doing things periodically instead of in real time though

**Comment by nisdas** [comment]
@prestonvanloon good point, we might not need to actually run this in a loop to trim it. It can be done in real time too, I can add it in a follow up PR

**Pull request merged**